File: patterns/_tests/BehavioralPatterns/Interpreter/cpp/integration.test.cpp
// ============================================================
//  Integration Tests — Interpreter: Arithmetic Expression Evaluator (C++)
//
//  Compile: g++ -std=c++17 -DTESTING -o integration_test integration.test.cpp \
//               ../../../../BehavioralPatterns/Interpreter/cpp/cpp.cpp
//  Run: ./integration_test
// ============================================================

#define TESTING
#include "../../../../BehavioralPatterns/Interpreter/cpp/cpp.cpp"
#include <cassert>
#include <iostream>
#include <stdexcept>
#include <string>
#include <cmath>

static int passed = 0;
static int failed = 0;

#define TEST(name, body) \
    do { \
        try { body; ++passed; std::cout << "  PASS  " << name << "\n"; } \
        catch (const std::exception& e) { ++failed; std::cout << "  FAIL  " << name << " : " << e.what() << "\n"; } \
        catch (...) { ++failed; std::cout << "  FAIL  " << name << " : unknown exception\n"; } \
    } while (false)

#define ASSERT_NEAR(a, b) do { if (std::abs((a) - (b)) > 0.001) throw std::runtime_error("Values not close"); } while(false)
#define ASSERT_TRUE(cond) do { if (!(cond)) throw std::runtime_error("Assertion failed: " #cond); } while(false)
#define ASSERT_THROWS(expr, ExType) \
    do { bool threw = false; try { expr; } catch (const ExType&) { threw = true; } \
         if (!threw) throw std::runtime_error("Expected exception not thrown"); } while(false)


int main() {
    std::cout << "\n  Integration Tests — Interpreter (C++)\n";
    std::cout << "  " << std::string(50, '-') << "\n\n";

    // ── Full expression evaluation end-to-end ─────────────────
    TEST("Three-level nested tree evaluates correctly", {
        // ((a + b) * (a - b)) / 5  with a=10, b=5  →  15
        auto a = var("a"); auto b = var("b");
        Context ctx; ctx.set("a", 10); ctx.set("b", 5);
        auto expr = div_(mul(add(a, b), sub(a, b)), num(5));
        ASSERT_NEAR(expr->interpret(ctx), 15.0);
    });

    TEST("Tree reuse: three different contexts", {
        auto a = var("a"); auto b = var("b");
        auto expr = mul(add(a, b), sub(a, b));

        struct TC { double av, bv, expected; };
        for (auto [av, bv, expected] : std::initializer_list<TC>{{10,5,75},{6,2,32},{3,1,8}}) {
            Context ctx; ctx.set("a", av); ctx.set("b", bv);
            ASSERT_NEAR(expr->interpret(ctx), expected);
        }
    });

    TEST("Display produces fully parenthesised string", {
        auto expr = mul(add(var("a"), var("b")), sub(var("a"), var("b")));
        std::string s = expr->display();
        ASSERT_TRUE(s.find("(a + b)") != std::string::npos);
        ASSERT_TRUE(s.find("(a - b)") != std::string::npos);
        ASSERT_TRUE(s[0] == '(');
    });

    // ── Error propagation ─────────────────────────────────────
    TEST("Missing variable in deep subtree surfaces", {
        auto expr = add(num(1), mul(var("missing"), num(2)));
        Context ctx;
        ASSERT_THROWS(expr->interpret(ctx), std::runtime_error);
    });

    TEST("Division by zero in subtree surfaces", {
        auto expr = add(num(1), div_(num(4), num(0)));
        Context ctx;
        ASSERT_THROWS(expr->interpret(ctx), std::runtime_error);
    });

    // ── Context independence ───────────────────────────────────
    TEST("Parallel evaluations do not interfere", {
        auto expr = add(var("a"), var("b"));
        Context ctx1; ctx1.set("a", 1);   ctx1.set("b", 2);
        Context ctx2; ctx2.set("a", 100); ctx2.set("b", 200);
        ASSERT_NEAR(expr->interpret(ctx1),   3.0);
        ASSERT_NEAR(expr->interpret(ctx2), 300.0);
    });

    // ── Arithmetic correctness ─────────────────────────────────
    TEST("Addition is commutative", {
        Context ctx; ctx.set("a", 3); ctx.set("b", 7);
        auto ab = add(var("a"), var("b"));
        auto ba = add(var("b"), var("a"));
        ASSERT_NEAR(ab->interpret(ctx), ba->interpret(ctx));
    });

    TEST("Multiply by zero gives zero", {
        Context ctx; ctx.set("a", 999);
        ASSERT_NEAR(mul(var("a"), num(0))->interpret(ctx), 0.0);
    });

    TEST("Divide by one gives same value", {
        Context ctx; ctx.set("a", 42);
        ASSERT_NEAR(div_(var("a"), num(1))->interpret(ctx), 42.0);
    });

    // ── Summary ───────────────────────────────────────────────
    std::cout << "\n  " << std::string(50, '-') << "\n";
    std::cout << "  " << passed << " passed, " << failed << " failed\n\n";
    return failed > 0 ? 1 : 0;
}

File: patterns/_tests/BehavioralPatterns/Interpreter/typescript/integration.test.ts
// ============================================================
//  Integration Tests — Interpreter: Arithmetic Expression Evaluator (TypeScript)
//
//  Run: npx mocha --require ts-node/register integration.test.ts
// ============================================================

import { describe, it } from 'mocha';
import assert from 'assert';
import { captureOutput } from '../../../_helpers/typescript/helpers';
import {
  Context,
  NumberExpression,
  VariableExpression,
  AddExpression,
  SubtractExpression,
  MultiplyExpression,
  DivideExpression,
} from '../../../../BehavioralPatterns/Interpreter/typescript/typescript';

// ── Full expression evaluation end-to-end ────────────────────
describe('End-to-end expression evaluation', () => {
  it('evaluates a three-level nested tree correctly', () => {
    // ((a + b) * (a - b)) / 5  with a=10, b=5  →  75/5 = 15
    const a   = new VariableExpression('a');
    const b   = new VariableExpression('b');
    const ctx = new Context().set('a', 10).set('b', 5);

    const expr = new DivideExpression(
      new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b)),
      new NumberExpression(5),
    );
    assert.strictEqual(expr.interpret(ctx), 15);
  });

  it('tree reuse: same tree, three different contexts', () => {
    const a    = new VariableExpression('a');
    const b    = new VariableExpression('b');
    const expr = new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b));

    const results = [
      { a: 10, b: 5,  expected: 75 },
      { a: 6,  b: 2,  expected: 32 },
      { a: 3,  b: 1,  expected: 8  },
    ];

    for (const { a: av, b: bv, expected } of results) {
      const ctx = new Context().set('a', av).set('b', bv);
      assert.strictEqual(expr.interpret(ctx), expected);
    }
  });

  it('toString() produces a fully parenthesised string', () => {
    const expr = new MultiplyExpression(
      new AddExpression(new VariableExpression('a'), new VariableExpression('b')),
      new SubtractExpression(new VariableExpression('a'), new VariableExpression('b')),
    );
    const s = expr.toString();
    assert.ok(s.includes('(a + b)'));
    assert.ok(s.includes('(a - b)'));
    assert.ok(s.startsWith('('));
  });
});


// ── Error propagation ─────────────────────────────────────────
describe('Error propagation through tree', () => {
  it('missing variable in deep sub-tree surfaces the error', () => {
    const expr = new AddExpression(
      new NumberExpression(1),
      new MultiplyExpression(new VariableExpression('missing'), new NumberExpression(2)),
    );
    assert.throws(() => expr.interpret(new Context()), /Undefined variable/);
  });

  it('division by zero in sub-tree surfaces the error', () => {
    const expr = new AddExpression(
      new NumberExpression(1),
      new DivideExpression(new NumberExpression(4), new NumberExpression(0)),
    );
    assert.throws(() => expr.interpret(new Context()), /Division by zero/);
  });
});


// ── Context independence ──────────────────────────────────────
describe('Context independence', () => {
  it('two parallel evaluations with different contexts do not interfere', () => {
    const a    = new VariableExpression('a');
    const b    = new VariableExpression('b');
    const expr = new AddExpression(a, b);

    const ctx1 = new Context().set('a', 1).set('b', 2);
    const ctx2 = new Context().set('a', 100).set('b', 200);

    assert.strictEqual(expr.interpret(ctx1), 3);
    assert.strictEqual(expr.interpret(ctx2), 300);
  });
});


// ── Commutative and associativity checks ─────────────────────
describe('Arithmetic correctness', () => {
  it('addition is commutative', () => {
    const ctx  = new Context().set('a', 3).set('b', 7);
    const ab   = new AddExpression(new VariableExpression('a'), new VariableExpression('b'));
    const ba   = new AddExpression(new VariableExpression('b'), new VariableExpression('a'));
    assert.strictEqual(ab.interpret(ctx), ba.interpret(ctx));
  });

  it('subtraction is not commutative', () => {
    const ctx  = new Context().set('a', 10).set('b', 3);
    const ab   = new SubtractExpression(new VariableExpression('a'), new VariableExpression('b'));
    const ba   = new SubtractExpression(new VariableExpression('b'), new VariableExpression('a'));
    assert.notStrictEqual(ab.interpret(ctx), ba.interpret(ctx));
  });

  it('multiply by zero always gives zero', () => {
    const ctx  = new Context().set('a', 999);
    const expr = new MultiplyExpression(new VariableExpression('a'), new NumberExpression(0));
    assert.strictEqual(expr.interpret(ctx), 0);
  });
});
